internal/controller/pod_controller_test.go

```diff
@@ -0,0 +1,265 @@
+			g.Expect(pod.Spec.SchedulingGates).NotTo(ConsistOf(corev1.PodSchedulingGate{Name: constants.PodSchedulingGateName}))
+		}).Should(Succeed())
+
+		pod = createNewPod(1, testName)
+		delete(pod.Annotations, constants.CatGateImagesHashAnnotation)
+		err = k8sClient.Update(ctx, pod)
+		Expect(err).NotTo(HaveOccurred())
+
+		Eventually(func(g Gomega) {
+			err = k8sClient.Get(ctx, client.ObjectKey{Name: fmt.Sprintf("%s-pod-%d", testName, 1), Namespace: testName}, pod)
+			g.Expect(err).NotTo(HaveOccurred())
+			g.Expect(pod.Spec.SchedulingGates).NotTo(ConsistOf(corev1.PodSchedulingGate{Name: constants.PodSchedulingGateName}))
+		}).Should(Succeed())
+	})
+
+	It("should limit number of schedulable pods based on status", func() {
+		testName := "multiple-pods-with-some-running-pods"
+		namespace := &corev1.Namespace{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: testName,
+			},
+		}
+		err := k8sClient.Create(ctx, namespace)
+		Expect(err).NotTo(HaveOccurred())
+
+		for i := 0; i < 8; i++ {
+			createNewPod(i, testName)
+		}
+		pods := &corev1.PodList{}
+
+		Eventually(func(g Gomega) {
+			err := k8sClient.List(ctx, pods, &client.ListOptions{Namespace: testName})
+			g.Expect(err).NotTo(HaveOccurred())
+			numSchedulable := 0
+			for _, pod := range pods.Items {
+				if !existsSchedulingGate(&pod) {
+					numSchedulable += 1
+				}
+			}
+			g.Expect(numSchedulable).To(Equal(1))
+		}).Should(Succeed())
+		updateStatusForPodWithSchedulingGate(testName)
+
+		Eventually(func(g Gomega) {
+			err := k8sClient.List(ctx, pods, &client.ListOptions{Namespace: testName})
+			g.Expect(err).NotTo(HaveOccurred())
+			numSchedulable := 0
+			for _, pod := range pods.Items {
+				if !existsSchedulingGate(&pod) {
+					numSchedulable += 1
+				}
+			}
+			g.Expect(numSchedulable).To(Equal(3))
+		}).Should(Succeed())
+		updateStatusForOnePodWithSchedulingGate(testName)
+
+		Eventually(func(g Gomega) {
+			err := k8sClient.List(ctx, pods, &client.ListOptions{Namespace: testName})
+			g.Expect(err).NotTo(HaveOccurred())
+			numSchedulable := 0
+			for _, pod := range pods.Items {
+				if !existsSchedulingGate(&pod) {
+					numSchedulable += 1
+				}
+			}
+			g.Expect(numSchedulable).To(Equal(6))
+		}).Should(Succeed())
+	})
+})
+
+func createNewPod(index int, testName string) *corev1.Pod {
+	newPod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: testName,
+			Name:      fmt.Sprintf("%s-pod-%d", testName, index),
+		},
+		Spec: corev1.PodSpec{
+			InitContainers: []corev1.Container{
+				{
+					Name:  "sample1",
+					Image: testName + ".example.com/sample1-image:1.0.0",
+				},
+			},
+			Containers: []corev1.Container{
+				{
+					Name:  "sample2",
+					Image: testName + ".example.com/sample2-image:1.0.0",
+				},
+			},
+			Hostname: fmt.Sprintf("%s-hostname-%d", testName, index),
+		},
+	}
+	err := k8sClient.Create(ctx, newPod)
+	Expect(err).NotTo(HaveOccurred())
+	return newPod
+}
+func updateStatusForPodWithSchedulingGate(namespace string) {
```

Comment:
Please add a line between functions.
Also, this function name is hard to understand because it updates statuses of the pods that don't have scheduling gates.

```suggestion
}

func updateStatusForPodWithSchedulingGate(namespace string) {
```